### ISSUE TITLE
Can't remove keys defined in values.yml

### ISSUE THREAD

**Issue by jagregory** [kubernetes/helm#1966]
Since the introduction of deep merging (#1620), it's now not possible to remove keys from `values.yml` entirely.

For example, the [telegraf values](https://github.com/kubernetes/charts/blob/master/stable/telegraf/values.yaml) has a default entry for `single.config.inputs.influxdb` (see: [L340](https://github.com/kubernetes/charts/blob/master/stable/telegraf/values.yaml#L340)). There's no way to completely remove this key through `--set` or a override file.

**Telegraf values.yml**

```yml
single:
  config:
    inputs:
      influxdb:
        urls:
          - "http://influxdb-influxdb.tick:8086/debug/vars"
        timeout: "5s"
```

**My telegraf-overrides.yml**

```yaml
single:
  config:
    inputs:
      influxdb: null # or something?
```

**Comment by technosophos** [comment]
Yeah, this appears to be true. Any suggestions about how we might support that?

**Comment by jagregory** [comment]
A few options come to mind:

  * A flag to indicate the values you're supplying *replace* the values from the chart. `helm install -f my-values.yml --overwrite`
  * A special value in the yaml to indicate not to merge, perhaps a `null`. e.g. *Null means delete this key*

 **Meta:** the reason this is a problem is because Telegraf doesn't have an explicit `enabled` property you can set, so the presence of a value is significant.

We could work around this by adding our own `enabled` key under `influxdb`, and check for the presence that in the templating and not write out the node; but this puts the onus on each chart to get it right.

**More meta:** I've also been thinking about whether we should be trying to generate config files from YAML in the first place. Several charts (the Influx stack in particular) allow you to specify configs in YAML which get translated into TOML, this translation is quite complicated and I suspect we're losing some information in translation. Each chart is having to do this itself. It might be better to just not try to do this at all, have the ConfigMap contain non-generated TOML which you can just pass your own in wholesale. Most of the time I already have a config file I want to use, and translating it to YAML is arduous (though I appreciate being able to abstract away complicated configs).

**Comment by jagregory** [comment]
Prior art for my last little ramble: [grafana.ini](https://github.com/kubernetes/charts/blob/master/stable/grafana/values.yaml#L174)

**Comment by technosophos** [comment]
I wouldn't be opposed at all to an `--overwrite` flag. That makes sense.

Some work did make it into Helm 2.2.0 which will allow subcharts to be switched on and off based on conditions. But it will likely take the charts repository a while to catch up on those.

**Comment by scottrigby** [comment]
Hey guys, just coming to this now. We were just brainstorming about this problem in [Slack](https://kubernetes.slack.com/archives/C51E88VDG/p1499293488448683) Today. And just as I was about to create an issue, I looked and found this one 😄 

So I prefer the `null` option, over `--overwrite`, because that could change more than we need. The immediate problem I see with an `--overwrite` flag is that would change the deep nesting behavior of all the keys/values from #1656 (`helm install some/chart -f values1.yaml -f values2.yaml --set foo.bar=baz`. I would think we want to be able to target specific keys, otherwise lots of breakage may happen. Whereas we could just add one check for a `null` value during the Helm deep nesting logic, and in that case nix the key.

**Comment by jascott1** [comment]
I also like the `null` option and  using it with `helm init --set` would allow advanced users to  completely remove keys (e.g. the default nodeSelector that specifies linux) and have more control of the Tiller Deployment manifest.

**Comment by thomastaylor312** [comment]
@scottrigby So I was looking at the code as it stands without your PR. It looks like `~` indicates `null` in YAML. I was able to overwrite any key using that notation. However, with that said, it won't actually _delete_ the key, it will just set its value to `null`. So I think your PR is a better option because it actually deletes the key from the map

**Comment by technosophos** [comment]
Yeah, I too prefer an explicit null.

**Comment by technosophos** [comment]
If ya'll agree, could we add this to 2.6 under the "stability and security" heading?

**Comment by thomastaylor312** [comment]
Yes, I agree and already labeled the PR as such 😄

**Comment by thomastaylor312** [comment]
Closed in #2648
